api/envoy/api/v2/core/address.proto

```diff
@@ -74,6 +74,40 @@ message TcpKeepalive {
   google.protobuf.UInt32Value keepalive_interval = 3;
 }
 
+// Generic socket option message. This would be used to set socket options that
+// might not exist in upstream kernels or precompiled Envoy binaries.
+message SocketOption {
+  // An optional name to give this socket option
+  // No special meaning is assumed.
+  string readable_name = 1;
+  // Corresponding to the level value passed to setsockopt, such as IPPROTO_TCP
+  int32 level = 2;
+  // The numeric name as passed to setsockopt
+  int32 name = 3;
+  // The setsockopt option value. Integers should be encoded in network byte order.
+  oneof value {
+    option (validate.required) = true;
+
+    // Because many sockopts take an int value.
+    int32 int = 4;
+    // Otherwise it's a byte buffer.
+    bytes buf = 5;
+  }
+  enum SocketState {
+    option (gogoproto.goproto_enum_prefix) = false;
+    // Socket options are applied after socket creation but before binding the socket to a port
+    STATE_PREBIND = 0;
+    // Socket options are applied after binding the socket to a port but before calling listen()
+    STATE_BOUND = 1;
+    // Socket options are applied after calling listen()
+    STATE_LISTENING = 2;
+  }
+  // For listeners the option can only be used on sockets in the given state.
+  // For BindConfig, state is ignored.
```

Comment:
It's required but ignored?

Reply:
If it's only required in one context and not the other, can we re-structure by making another message that is SocketOptionWithSocketState, and remove it from here, so that it's only present in the contexts where it is useful?